### PR TITLE
Feature/narwhalify median imputer

### PR DESCRIPTION
**Pull request by Chip2916**
As per issue #317 

One weirdness around applying .median() method in the transformer which means I've had to add a workaround (which I don't like at all). I've commented out the solution I expected to work [here](https://github.com/lvgig/tubular/commit/b198561ca5dbdfc341a4fc53d52b2e737ead64b2#diff-49429640bd11a61fef0ea2bcb010bfb24ce43e4804c5b9d1f09f744c7e7ae48cR247). Would appreciate feedback on this, works fine if the .median() method is replaced with .mode() / .mean() so wondering if it's a bug. I suspect I'm just missing something, but happy to raise bug if not.

### CODE REVIEW

File: tubular/imputers.py [hunk 1]
```diff
@@ -213,30 +214,43 @@ def fit(self, X: pd.DataFrame, y: pd.Series | None = None) -> pd.DataFrame:
...
-                self.impute_values_[c] = X[c].median()
+                filtered = X.filter(~nw.col(c).is_null())
+
+                # need to return None not NaN
+                if len(filtered) <= 0:
+                    median = None
+
+                else:
+                    # why doesn't this work - median = X.select(nw.col(c).median()).item()
+                    # potential bug?
```
[the remark this answers is not in the document]
Comment (Chip2916) [reply tubular/imputers.py:248]: Hi Marco, not at all, appreciate the feedback!

Bit of a head scratcher - the full traceback is below. For some reason when I use "median = X.select(nw.col(c).median()).item()" I get the "AttributeError: 'Expr' object has no attribute 'median'" error.

Replacing the .median() method with .mean() throws no issue. As per the Polars documentation [here](https://docs.pola.rs/api/python/dev/reference/expressions/api/polars.Expr.median.html) median is an accepted expr method. Have also checked that 'X' is a dataframe when trying to obtain the median.
___________________________________________________________________________________ TestTransform.test_pandas_index_not_updated[pandas] ___________________________________________________________________________________

self = <test_MedianImputer.TestTransform object at 0x75917c699610>
initialized_transformers = {'ArbitraryImputer': ArbitraryImputer(columns=['b'], impute_value=1), 'BaseCappingTransformer': BaseCappingTransformer...eCrossColumnNumericTransformer(adjust_column='b',
                                  mappings={'a': {1: 2, 3: 4}}), ...}
minimal_dataframe_lookup = {'ArbitraryImputer':         a      b      c      d      e
0   34.48  12.03  17.06  25.94  94.30
2   21.71  20.32  12....14  75.70  59.46  20.59   4.60  30.79
16  14.08  11.02   3.82  67.13  66.19
18  61.31  60.68  39.73  10.38  25.44, ...}

    @pytest.mark.parametrize(
        "minimal_dataframe_lookup",
        ["pandas"],
        indirect=True,
    )
    def test_pandas_index_not_updated(
        self,
        initialized_transformers,
        minimal_dataframe_lookup,
    ):
        """Test that the original (pandas) dataframe index is not transformed when transform method used."""
    
        df = minimal_dataframe_lookup[self.transformer_name]
        x = initialized_transformers[self.transformer_name]
    
        # update to abnormal index
        df.index = [2 * i for i in df.index]
    
        original_df = copy.deepcopy(df)
    
>       x = x.fit(df, df["a"])

tests/base_tests.py:866: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/vscode/.local/lib/python3.9/site-packages/narwhals/translate.py:766: in wrapper
    result = func(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = MedianImputer(columns=['b']), X = ┌───────────────────────────────────────┐
| Narwhals DataFrame                    |
| Use `.to_native` to see native output |
└───────────────────────────────────────┘
y = ┌─────────────────────────────────────────┐
| Narwhals Series                         |
| Use `.to_native()` to see native output |
└─────────────────────────────────────────┘

    @nw.narwhalify
    def fit(self, X: FrameT, y: nw.Series | None = None) -> FrameT:
        """Calculate median values to impute with from X.
    
        Parameters
        ----------
        X : pd/pl.DataFrame
            Data to "learn" the median values from.
    
        y : None or pd/pl.Series, default = None
            Not required.
    
        """
        super().fit(X, y)
    
        self.impute_values_ = {}
    
        if self.weights_column is not None:
            WeightColumnMixin.check_weights_column(self, X, self.weights_column)
    
            for c in self.columns:
                # filter out null rows so their weight doesn't influence calc
                filtered = X.filter(~nw.col(c).is_null())
    
                # below algorithm only works for >1 non null values
                if len(filtered) <= 0:
                    median = None
    
                else:
                    # first sort df by column to be imputed (order of weight column shouldn't matter for median)
                    filtered = filtered.sort(c)
    
                    # next calculate cumulative weight sums
                    cumsum = filtered[self.weights_column].cum_sum()
    
                    # find midpoint
                    cutoff = filtered[self.weights_column].sum() / 2.0
    
                    # find first value >= this point
                    median = filtered.filter(cumsum >= cutoff).select(c)[0].item()
    
                self.impute_values_[c] = median
    
        else:
            for c in self.columns:
                filtered = X.filter(~nw.col(c).is_null())
    
                # need to return None not NaN
                if len(filtered) <= 0:
                    median = None
    
                else:
                    # why doesn't this work - median = X.select(nw.col(c).mean()).item()
                    # potential bug?
>                   median = X.select(nw.col(c).median()).item()
E                   AttributeError: 'Expr' object has no attribute 'median'

tubular/imputers.py:249: AttributeError

[the remark this answers is not in the document]
Comment (Chip2916) [reply tubular/imputers.py]: Hey @MarcoGorelli - apologies for not sharing the full traceback - copy and paste error! Either way, you were right (and I shoulda checked the docs...), it just needed updating to a newer version of narwhals. We appreciate your continued support on Tubular!

File: tests/base_tests.py [hunk 2]
```diff
@@ -585,7 +585,7 @@ def test_fit_not_changing_data_weighted(
     @pytest.mark.parametrize(
         "bad_weight_value, expected_message",
         [
-            (np.nan, "weight column must be non-null"),
+            (None, "weight column must be non-null"),
```
[the remark this answers is not in the document]
Comment (Chip2916) [reply tests/base_tests.py]: As per discussions with yourself and Liam - I've pulled Liam's temporary fix into my branch which has allowed the unblocking of tests for np.nan. Also seen nw has been upgraded to much newer version (presumably to double check that is_nan hadn't been added) - for this and the below comment below I think this brings this in line with agreed method. Will look to raise a separate issue around is_nan method.

File: tubular/imputers.py [hunk 3]
```diff
@@ -192,51 +192,52 @@ def __init__(
...
-        if self.weights_column is not None:
-            WeightColumnMixin.check_weights_column(self, X, self.weights_column)
+        for c in self.columns:
+            # filter out null rows so their weight doesn't influence calc
+            filtered = X.filter(~nw.col(c).is_null())
 
-            for c in self.columns:
-                # filter out null rows so their weight doesn't influence calc
-                filtered = X[X[c].notna()]
+            # if column is only nulls, then median is None
```
[the remark this answers is not in the document]
Comment (Chip2916) [reply tubular/imputers.py:218]: As above!

**Review by davidhopkinson26 — APPROVED** [review]
Thanks @Chip2916

**Pull request merged**